File: javascript/two_pointers/1768. Merge Strings Alternately.js
/**
 * @param {string} word1
 * @param {string} word2
 * @return {string}
 */

// own (modified)
var mergeAlternately = function(word1, word2) {
    let res = "";

    for (let i=0; i < Math.max(word1.length, word2.length); i++) {
        if (i < word1.length) res += word1[i];
        if (i < word2.length) res += word2[i];
    }
    return res;
};

// Two pointer
var mergeAlternately = function(word1, word2) {
    let len1 = word1.length, len2 = word2.length;
    let i = j = 0;
    let res = "";

    while (i < len1 || j < len2) {
        if (i < len1) {
            res += word1[i];
            i++;
        }
        if (j < len2) {
            res += word2[j];
            j++;
        }
    }
    return res;
};

File: python/two_pointers/1768. Merge Strings Alternately.py
'''
own; Two pointer:
Time: O(n)
Space: O(n)
'''

class Solution:
    # own
    def mergeAlternately(self, word1: str, word2: str) -> str:
        res = ""
        for i in range(min(len(word1), len(word2))):
            res += word1[i] + word2[i]

        for j in range(i, max(len(word1), len(word2))):
            res += word1[j] if (len(word1) > len(word2)) else word2[j]

        return res
    
    # Two pointer
    def mergeAlternately(self, word1: str, word2: str) -> str:
        res = ""
        i = j = 0
        len1, len2 = len(word1), len(word2)
        while i < len1 or j < len2:
            if i < len1:
                res += word1[i]
                i += 1
            if j < len2:
                res += word2[j]
                j += 1

        return res